fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java

```diff
@@ -0,0 +1,458 @@
+            // only support distinct aggregation push down window
+            if (!context.origAggregator.getAggregations().isEmpty() || !windowOp.canPushDownAgg()) {
+                return visit(optExpression, context);
+            }
+
+            List<ColumnRefOperator> additionalGroupBys = Lists.newArrayList();
+            windowOp.getPartitionExpressions().stream()
+                    .flatMap(e -> e.getUsedColumns().getStream().mapToObj(factory::getColumnRef)).forEach(
+                            additionalGroupBys::add);
+
+            windowOp.getWindowCall().keySet().forEach(context.groupBys::remove);
+            windowOp.getOrderByElements().forEach(o -> additionalGroupBys.add(o.getColumnRef()));
+            additionalGroupBys.forEach(c -> context.groupBys.put(c, c));
+            context.aggregations.putAll(windowOp.getWindowCall());
+            context.hasWindow = true;
+            return context;
+        }
+
+        @Override
+        public AggregatePushDownContext visitLogicalProject(OptExpression optExpression,
+                                                            AggregatePushDownContext context) {
+            if (canNotPushDown(optExpression, context)) {
+                return visit(optExpression, context);
+            }
+
+            LogicalProjectOperator projectOp = optExpression.getOp().cast();
+
+            if (projectOp.getColumnRefMap().entrySet().stream().allMatch(e -> e.getValue().equals(e.getKey()))) {
+                return context;
+            }
+
+            // rewrite
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(projectOp.getColumnRefMap());
+            context.aggregations.replaceAll((k, v) -> (CallOperator) rewriter.rewrite(v));
+            context.groupBys.replaceAll((k, v) -> rewriter.rewrite(v));
+
+            if (projectOp.getColumnRefMap().values().stream().allMatch(ScalarOperator::isColumnRef)) {
+                return context;
+            }
+
+            if (!context.aggregations.isEmpty() &&
+                    !context.aggregations.values().stream().allMatch(c -> c.getChildren().stream().allMatch(
+                            ScalarOperator::isColumnRef))) {
+                return visit(optExpression, context);
+            }
+            return context;
+        }
+
+        @Override
+        public AggregatePushDownContext visitLogicalFilter(OptExpression optExpression,
+                                                           AggregatePushDownContext context) {
+            if (canNotPushDown(optExpression, context)) {
+                return visit(optExpression, context);
+            }
+
+            // add filter columns in groupBys
+            LogicalFilterOperator filter = optExpression.getOp().cast();
+            filter.getRequiredChildInputColumns().getStream().mapToObj(factory::getColumnRef)
+                    .forEach(v -> context.groupBys.put(v, v));
```

Comment:
Why add filter columns into `groupBys`, this filter is above the window or below the window?

Reply:
commented added